### ISSUE TITLE
error: internal compiler error: /checkout/src/librustc_borrowck/borrowck/mod.rs:808: aliasability violation for static `cannot borrow data mutably`

### ISSUE THREAD

**Issue by Ekleog** [rust-lang/rust#42344]
When compiling the following code, with rustc (nightly, beta or stable at the moment):

```rust
static TAB: [&mut [u8]; 0] = [];
pub unsafe fn test() {
    TAB[0].iter_mut();
}
pub fn main() {}
```

I get the following error:

```
error: internal compiler error: /checkout/src/librustc_borrowck/borrowck/mod.rs:808: aliasability violation for static `cannot borrow data mutably`
```

Here is the relevant playground link: https://is.gd/42slwV

Hope that helps,
Leo

**Comment by kennytm** [comment]
E0388 strikes back. Regression from 1.16.0 (introduced by #39927). Error in 1.16.0:

```
error[E0388]: cannot borrow data mutably in a static location
 --> <source>:3:5
  |
3 |     TAB[0].iter_mut();
  |     ^^^^^^ cannot write data in a static definition
error: aborting due to previous error
Compiler exited with result code 101
```

**Comment by Ekleog** [comment]
As my example may lead to confusion, this also happens with arrays of non-zero size: https://is.gd/lE9Th8

```rust
static FOO: [u8; 1] = [0];
static TAB: [&mut [u8]; 1] = [&mut FOO];
pub unsafe fn test() {
    TAB[0].iter_mut();
}
pub fn main() {}
```

**Comment by nikomatsakis** [comment]
([Works properly with NLL enabed.](https://play.rust-lang.org/?gist=94cf886694c0800b1c25224358ba96f4&version=nightly))

**Comment by Ekleog** [comment]
Hmm, I don't think this commit fixes the underlying issue? It does fix the non-zero-size-array case, but not the 0-size one, so I guess it only makes it harder to actually build this [&mut[u8]; _] static type? (Or maybe it just didn't land in rust playground yet? but as the non-empty-array case is fixed with nightly on playground, I guess it did land)

tl;dr: this still ICE's on playground's nightly (with NLL disabled)
```rust
static TAB: [&mut [u8]; 0] = [];
pub unsafe fn test() {
    TAB[0].iter_mut();
}
pub fn main() {}
```

**Comment by kennytm** [comment]
@Ekleog It has just been merged 25 minutes ago... The nightly won't be released until 14 more hours. Please check again tomorrow 🙂.

**Comment by Ekleog** [comment]
Oh i misread both bors' messages (thought it was merged 16 hours ago, but I guess the merge failed) and the PR title… indeed as this example is added as a test it can only work once it's released!

Sorry for the noise and thanks for the fix!
